.pre-commit-config.yaml

```diff
@@ -38,6 +39,25 @@ repos:
       - id: numpydoc-validation
         files: ^pyvista/
 
+  - repo: https://github.com/pyvista/pre-commit-mypy
+    rev: v0.1.0
+    hooks:
+      - id: mypy
+        args: [] # Args should be blank. Set args via 'pyproject.toml' instead.
+        additional_dependencies:
+          # These should match `pyvista[typing]` dependencies in 'pyproject.toml'
+          [
+            "vtk<9.4.0",
```

Comment:
How will the versions here be kept up to date with those in pyproject?  I think dependabot will do one, and precommit.ci will do the other.  These will always be out of sync.

We can also consider the second suggestion in https://github.com/python/mypy/issues/13916 to

>Use language: system (and maybe a helper script) to run something from outside a pre-commit environment

Reply:
There are new parametrized tests included with this PR (see tests/typing). These tests compare dependencies in pyproject.toml with those in pre-commit-config.yaml. The tests fail if the dependencies are out of sync. 

So, when dependabot bumps a dependency in pyproject.toml, there will initially be a test failure, and the pre-commit config will need to be manually bumped to resolve the failure. This should keep everything in sync.

Reply:
> We can also consider the second suggestion in https://github.com/python/mypy/issues/13916 to
> 
> 
> 
> >Use language: system (and maybe a helper script) to run something from outside a pre-commit environment

I considered this option, but I think it's good to keep typing dependencies separate from other dependencies (i.e. test, docs, and dev). And so the problem with using system is that now devs will need to make sure they do `pip install -e .[typing]` before running pre-commit or else they risk having the wrong dependencies installed. 

For example, I found that doing `pip install -e .[dev]` will install numpy 1.26, because some dependencies (e.g. osmnx) have numpy pinned less than 2.0. This is not compatible with the typing environment where we want numpy>=2.0. 

So, to avoid this kind of complexity (and possible frustration for devs), I like the idea of letting pre-commit configure the venv.

Reply:
> We can also consider the second suggestion in [python/mypy#13916](https://github.com/python/mypy/issues/13916) to
> 
> > Use language: system (and maybe a helper script) to run something from outside a pre-commit environment

If there is a strong preference for using `system` let me know and I will rework this PR accordingly

Reply:
I do not have a strong preference here. I would ask @tkoyama010 about the maintenance burden as they have been taking the lead on version updates recently.

While the addition of the test, which I haven't thoroughly reviewed, should enhance the robustness of keeping the requirements in sync, this is now code that will also need to be maintained, also adding several new test dependencies, etc..

Reply:
The extra maintenance burden is an important consideration. There doesn't seem to be a perfect solution. 

If we use `system`, the burden of updating the environment is now placed on the developer. In this case it could lead to local errors not showing up in CI or vice-versa. This could be unexpected and may be difficult to track down, especially if a typing dependency is bumped mid-way through PR changes.

Perhaps it's better not to use `mypy` in pre-commit at all, and rely solely on CI for this. Devs may then optionally choose to run `mypy` locally if desired between commits. We could add a note in `CONTRIBUTING.md` about this.

Reply:
> Perhaps it's better not to use mypy in pre-commit at all, and rely solely on CI for this. Devs may then optionally choose to run mypy locally if desired between commits. We could add a note in CONTRIBUTING.md about this.

This is my personal preference, but it is only that.  I think other devs should weigh in on it.  There are certainly upsides to the approach in this PR.